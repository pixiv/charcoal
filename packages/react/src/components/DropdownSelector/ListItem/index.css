.charcoal-list-item {
  list-style: none;
  display: flex;
  align-items: center;
  min-height: 40px;
  cursor: pointer;
  outline: none;

  padding-right: 16px;
  padding-left: 16px;

  transition: background-color 0.2s;
}

.charcoal-list-item:not([aria-disabled='true']):hover,
.charcoal-list-item:not([aria-disabled='true']):focus,
.charcoal-list-item:not([aria-disabled='true']):focus-within {
  background-color: var(--charcoal-color-container-secondary-default);
}

.charcoal-list-item[aria-disabled='true'] {
  cursor: default;
  opacity: 0.32;
}
