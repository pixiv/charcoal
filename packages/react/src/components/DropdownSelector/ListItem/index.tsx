import React, { ReactNode } from 'react'
import styled from 'styled-components'
import { theme } from '../../../styled'

export type CustomJSXElement =
  | keyof JSX.IntrinsicElements
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  | React.JSXElementConstructor<any>

export type ListItemProps<T extends CustomJSXElement = 'div'> = {
  children?: ReactNode
  as?: T
} & Omit<React.ComponentProps<T>, 'children'>

/**
 * リストのある要素を示すコンポーネント
 *
 * asを用いて拡張することができる
 * @example
 * ```
 * <ListItem as="a" href="#">Link</ListItem>
 * <ListItem as={NextLink} href="#">NextLink</ListItem>
 * ```
 */
export default function ListItem<T extends CustomJSXElement = 'div'>(
  props: ListItemProps<T>
) {
  const { children, ...rest } = props
  return (
    <StyledLi role="option">
      <ItemDiv {...rest}>
        {props.children}
      </ItemDiv>
    </StyledLi>
  )
}

const StyledLi = styled.li`
  list-style: none;
`

const ItemDiv = styled.div`
  display: flex;
  align-items: center;
  min-height: 40px;
  cursor: pointer;
  outline: none;

  ${theme((o) => [o.padding.horizontal(16), o.disabled])}

  &[aria-disabled="true"] {
    cursor: default;
  }

  :hover,
  :focus,
  :focus-within {
    ${theme((o) => [o.bg.surface3])}
  }
`
