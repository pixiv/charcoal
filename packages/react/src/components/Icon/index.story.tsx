import React from 'react'
import Icon, { IconProps } from '.'
import { KNOWN_ICON_FILES } from '@charcoal-ui/icons'
import { Story } from '../../_lib/compat'

export default {
  title: 'Icon',
  component: Icon,
  argTypes: {
    name: {
      control: {
        type: 'select',
        options: KNOWN_ICON_FILES,
      },
    },
    scale: {
      control: {
        type: 'select',
        options: [1, 2, 3],
      },
    },
  },
}

// NOTICE: コード生成のタイミングで不当に型エラーが出ることがあるので name を any にする
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const Template: Story<IconProps & { name: any }> = (props) => (
  <Icon {...props} />
)

export const Default = Template.bind({})

Default.args = { name: KNOWN_ICON_FILES[0], scale: 1 }
