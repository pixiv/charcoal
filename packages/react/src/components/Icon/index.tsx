import * as React from 'react'

import type { PixivIcon, Props } from '@charcoal-ui/icons'

export interface OwnProps {
  unsafeNonGuidelineScale?: number
  className?: string
}

export interface IconProps
  extends OwnProps,
    React.PropsWithoutRef<
      Omit<Props, 'class' | 'unsafe-non-guideline-scale' | 'css'>
    > {}

const Icon = React.forwardRef<PixivIcon, IconProps>(function IconInner(
  { name, scale, unsafeNonGuidelineScale, className, ...rest },
  ref
) {
  return (
    <pixiv-icon
      ref={ref}
      name={name}
      scale={scale}
      unsafe-non-guideline-scale={unsafeNonGuidelineScale}
      class={className}
      {...rest}
    />
  )
})

export default Icon
