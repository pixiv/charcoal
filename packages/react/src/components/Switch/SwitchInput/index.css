.charcoal-switch-input {
  cursor: pointer;
  appearance: none;
  display: inline-flex;
  position: relative;
  box-sizing: border-box;
  width: 28px;
  border: 2px solid transparent;

  transition-property: background-color, box-shadow;
  transition-duration: 0.2s;

  outline: none;
  border-radius: 16px;
  height: 16px;
  margin: 0;
  background-color: var(--charcoal-color-container-neutral-default);
}

.charcoal-switch-input:disabled,
.charcoal-switch-input[readonly] {
  cursor: default;
}

.charcoal-switch-input::after {
  content: '';
  position: absolute;
  display: block;
  top: 0;
  left: 0;
  width: 12px;
  height: 12px;
  transform: translateX(0);
  transition: transform 0.2s;
  border-radius: 1024px;
  background-color: var(--charcoal-color-icon-on-primary-default);
}

.charcoal-switch-input:checked::after {
  transform: translateX(12px);
  transition: transform 0.2s;
}

.charcoal-switch-input:checked {
  background-color: var(--charcoal-color-container-primary-default);
}

.charcoal-switch-input:not(:disabled):hover {
  background-color: var(--charcoal-color-container-neutral-hover);
}

.charcoal-switch-input:not(:disabled):active {
  background-color: var(--charcoal-color-container-neutral-press);
}

.charcoal-switch-input:not(:disabled):focus {
  outline: none;
  box-shadow: var(--charcoal-color-custom-focus-effect);
}
.charcoal-switch-input:not(:disabled):focus-visible {
  box-shadow: var(--charcoal-color-custom-focus-effect);
}
.charcoal-switch-input:not(:disabled):focus:not(:focus-visible) {
  box-shadow: none;
}

.charcoal-switch-input:not(:disabled)::after:hover {
  background-color: var(--charcoal-color-icon-on-primary-hover);
}

.charcoal-switch-input:not(:disabled)::after:active {
  background-color: var(--charcoal-color-icon-on-primary-press);
}

.charcoal-switch-input:not(:disabled):checked:hover {
  background-color: var(--charcoal-color-container-primary-hover);
}

.charcoal-switch-input:not(:disabled):checked:active {
  background-color: var(--charcoal-color-container-primary-press);
}
