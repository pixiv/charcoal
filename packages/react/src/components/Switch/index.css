.charcoal-switch__label {
  display: inline-grid;
  grid-template-columns: auto 1fr;
  align-items: center;
  cursor: pointer;
  outline: 0;
  gap: 4px;
}

.charcoal-switch__label[aria-disabled='true'] {
  cursor: default;
  opacity: 0.32;
}

.charcoal-switch__label_div {
  font-size: 14px;
  line-height: 22px;
  color: var(--charcoal-color-text-default);
}
