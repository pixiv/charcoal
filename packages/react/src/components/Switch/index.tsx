import { useSwitch } from '@react-aria/switch'
import type { AriaSwitchProps } from '@react-types/switch'
import React, { useRef, useMemo } from 'react'
import { useToggleState } from 'react-stately'
import styled from 'styled-components'
import { theme } from '../../styled'
import { disabledSelector } from '@charcoal-ui/utils'

export type SwitchProps = {
  name: string
  className?: string
  value?: string
  checked?: boolean
  disabled?: boolean
  onChange(checked: boolean): void
} & (
  | // children か label は片方が必須
  {
      children: React.ReactNode
    }
  | {
      label: string
    }
)

export default function SwitchCheckbox(props: SwitchProps) {
  const { disabled, className } = props

  const ariaSwitchProps: AriaSwitchProps = useMemo(
    () => ({
      ...props,

      // children がいない場合は aria-label をつけないといけない
      'aria-label': 'children' in props ? undefined : props.label,
      isDisabled: props.disabled,
      isSelected: props.checked,
    }),
    [props]
  )

  const state = useToggleState(ariaSwitchProps)
  const ref = useRef<HTMLInputElement>(null)
  const {
    inputProps: { className: _className, type: _type, ...rest },
  } = useSwitch(ariaSwitchProps, state, ref)

  return (
    <Label className={className} aria-disabled={disabled}>
      <SwitchInput {...rest} ref={ref} />
      {'children' in props ? (
        // eslint-disable-next-line react/destructuring-assignment
        <LabelInner>{props.children}</LabelInner>
      ) : undefined}
    </Label>
  )
}

const Label = styled.label`
  display: inline-grid;
  grid-template-columns: auto 1fr;
  align-items: center;
  cursor: pointer;
  outline: 0;

  ${theme((o) => o.disabled)}

  ${disabledSelector} {
    cursor: default;
  }
`

const LabelInner = styled.div`
  ${theme((o) => [
    o.typography(14).preserveHalfLeading,
    o.font.text2,
    o.margin.left(4),
  ])}
`

const SwitchInput = styled.input.attrs({
  type: 'checkbox',
})`
  &[type='checkbox'] {
    appearance: none;
    display: inline-flex;
    position: relative;
    box-sizing: border-box;
    width: 28px;
    border: 2px solid transparent;
    transition: box-shadow 0.2s, background-color 0.2s;
    cursor: inherit;
    ${theme((o) => [
      o.borderRadius(16),
      o.height.px(16),
      o.bg.text4.hover.press,
      o.outline.default.focus,
      o.margin.all(0),
    ])}

    &::after {
      content: '';
      position: absolute;
      display: block;
      top: 0;
      left: 0;
      width: 12px;
      height: 12px;
      transform: translateX(0);
      transition: transform 0.2s;
      ${theme((o) => [o.bg.text5.hover.press, o.borderRadius('oval')])}
    }

    &:checked {
      /* FIXME: o.outline.default.focus の transition に o.bg.brand の transition が打ち消されてしまう */
      transition: all 0.2s !important;
      ${theme((o) => o.bg.brand.hover.press)}

      &::after {
        transform: translateX(12px);
      }
    }
  }
`
