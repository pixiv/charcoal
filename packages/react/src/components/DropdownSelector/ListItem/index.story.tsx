import React, { useState } from 'react'
import { Story } from '../../../_lib/compat'
import Icon from '../../Icon'
import Switch from '../../Switch'
import ListItem, { ListItemProps } from '.'
import styled from 'styled-components'

export default {
  title: 'DropdownSelector/ListItem',
  component: ListItem,
}

const CustomLink = styled.a`
  color: red;
`

export const Basic: Story<ListItemProps> = () => {
  const [checked, setChecked] = useState(false)
  const handleCheck = () => {
    setChecked((v) => !v)
  }
  return (
    <>
      <ListItem>Item</ListItem>
      <ListItem>
        <Icon name="16/Add" /> Add
      </ListItem>
      <ListItem as="a" href="#">
        Normal Link
      </ListItem>
      <ListItem as={CustomLink} href="#">
        Custom Link
      </ListItem>
      <ListItem onClick={handleCheck}>
        Switch
        <div
          style={{
            marginLeft: 'auto',
          }}
        >
          <Switch
            label="hello"
            name="hello"
            onChange={handleCheck}
            checked={checked}
          />
        </div>
      </ListItem>
    </>
  )
}
