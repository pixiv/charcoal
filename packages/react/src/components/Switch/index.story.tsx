import { useState } from 'react'
import Switch from '.'
import { Meta, StoryObj } from '@storybook/react'

export default {
  title: 'Switch',
  component: Switch,
  parameters: {
    layout: 'centered',
  },
} as Meta<typeof Switch>

export const Default: StoryObj<typeof Switch> = {
  render: function Render(args) {
    const [checked, setChecked] = useState(false)
    return (
      <div>
        <Switch
          {...args}
          name="name"
          onChange={(v) => {
            setChecked(v)
          }}
          checked={args.checked ?? checked}
        />
      </div>
    )
  },
}

export const Checked: StoryObj<typeof Switch> = {
  render: function Render() {
    const [checked, setChecked] = useState(true)
    return (
      <div>
        <Switch
          name="name"
          label="checked"
          onChange={(v) => {
            setChecked(v)
          }}
          checked={checked}
        />
      </div>
    )
  },
}

export const Label: StoryObj<typeof Switch> = {
  render: function Render() {
    const [checked, setChecked] = useState(false)
    return (
      <div>
        <Switch name="label" checked={checked} onChange={setChecked}>
          Label
        </Switch>
      </div>
    )
  },
}

export const Disabled: StoryObj<typeof Switch> = {
  render: function Render() {
    const [checked, setChecked] = useState(false)
    return (
      <div>
        <Switch
          name="disabled"
          checked={checked}
          onChange={setChecked}
          disabled
        >
          Label
        </Switch>
      </div>
    )
  },
}
