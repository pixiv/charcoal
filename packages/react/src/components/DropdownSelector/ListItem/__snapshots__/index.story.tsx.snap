// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots DropdownSelector/ListItem Basic 1`] = `
.c3 {
  display: inline-grid;
  grid-template-columns: auto 1fr;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  outline: 0;
}

.c3:disabled,
.c3[aria-disabled]:not([aria-disabled='false']) {
  opacity: 0.32;
  cursor: default;
}

.c3:active > input {
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

.c4 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  box-sizing: border-box;
  width: 28px;
  border: 2px solid transparent;
  -webkit-transition-property: background-color,box-shadow;
  transition-property: background-color,box-shadow;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
  cursor: inherit;
  outline: none;
  border-radius: 16px;
  height: 16px;
  margin: 0;
  background-color: var(--charcoal-text4);
}

.c4:hover {
  background-color: var(--charcoal-text4-hover);
}

.c4:active {
  background-color: var(--charcoal-text4-press);
}

.c4:focus {
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

.c4::after {
  content: '';
  position: absolute;
  display: block;
  top: 0;
  left: 0;
  width: 12px;
  height: 12px;
  -webkit-transform: translateX(0);
  -ms-transform: translateX(0);
  transform: translateX(0);
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
  border-radius: 1024px;
  background-color: var(--charcoal-text5);
}

.c4::after:hover {
  background-color: var(--charcoal-text5-hover);
}

.c4::after:active {
  background-color: var(--charcoal-text5-press);
}

.c4:checked {
  background-color: var(--charcoal-brand);
}

.c4:checked:hover {
  background-color: var(--charcoal-brand-hover);
}

.c4:checked:active {
  background-color: var(--charcoal-brand-press);
}

.c4:checked::after {
  -webkit-transform: translateX(12px);
  -ms-transform: translateX(12px);
  transform: translateX(12px);
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
}

.c0 {
  list-style: none;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-height: 40px;
  cursor: pointer;
  outline: none;
  padding-right: 16px;
  padding-left: 16px;
}

.c1:disabled,
.c1[aria-disabled]:not([aria-disabled='false']) {
  opacity: 0.32;
  cursor: default;
}

.c1:hover,
.c1:focus,
.c1:focus-within {
  background-color: var(--charcoal-surface3);
}

.c2 {
  color: red;
}

<div
  data-dark={false}
>
  <li
    className="c0"
    role="option"
  >
    <div
      className="c1"
    >
      Item
    </div>
  </li>
  <li
    className="c0"
    role="option"
  >
    <div
      className="c1"
    >
      <pixiv-icon
        name="16/Add"
      />
       Add
    </div>
  </li>
  <li
    className="c0"
    role="option"
  >
    <a
      className="c1"
      href="#"
    >
      Normal Link
    </a>
  </li>
  <li
    className="c0"
    role="option"
  >
    <a
      className="c2 c1"
      href="#"
    >
      Custom Link
    </a>
  </li>
  <li
    className="c0"
    role="option"
  >
    <div
      className="c1"
      onClick={[Function]}
    >
      Switch
      <div
        style={
          Object {
            "marginLeft": "auto",
          }
        }
      >
        <label
          className="c3"
        >
          <input
            aria-label="hello"
            checked={false}
            className="c4"
            disabled={false}
            name="hello"
            onChange={[Function]}
            onClick={[Function]}
            onDragStart={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchCancel={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="switch"
            type="checkbox"
          />
        </label>
      </div>
    </div>
  </li>
</div>
`;
